RFC-0001-prsestissimo-metrics.md

```diff
@@ -0,0 +1,377 @@
+
+At any instant, we have a snapshot of metrics maintained which is overwritten on new updates.
+
+
+#### 
+**How to Export:**
+
+Prestissimo as exporter: To keep it simple, the prestissimo worker will behave like an exporter and expose the REST endpoint `/v1/info/metrics`. The Prometheus server can be configured to pull metrics from the worker itself. Again, it is up to the user if they want to introduce another layer in between the Time series DB and the worker.
+
+A sidecar metric exporter: A sidecar HTTP server that fetches metrics from Prestissimo. This would require launching as a child process of Prestissimo or as a separate container in the same POD as Prestissimo. Pros: In case of Prestissimo crash, we still have a way to get the last set of metrics from the node. To support graceful failure handling, we may have to dump metrics to disk and the exporter must read from disk when Prestissimo is down. Cons: Spawning a new process as a sidecar would take additional memory and CPU.
+
+Currently, we have only prototyped Prestissimo as an exporter. Both approaches can be built using the [prometheus-cpp](https://github.com/jupp0r/prometheus-cpp/tree/master) library.
+
+
+### 
+**Storing metric values in BaseStatsReporter:**
+
+
+#### 
+**Memory footprint**
+
+We have in total 103 + 28 + 7= 138 metric keys defined in PrestoCPP and Velox. Here is the current split by metric type:
+
+
+<table>
+  <tr>
+   <td><strong>Metric Type</strong>
+   </td>
+   <td><strong>PrestoCPP Count</strong>
+   </td>
+   <td><strong>Velox Count</strong>
+   </td>
+   <td><strong>Total</strong>
+   </td>
+  </tr>
+  <tr>
+   <td>COUNT
+   </td>
+   <td>3
+   </td>
+   <td>10
+   </td>
+   <td>13
+   </td>
+  </tr>
+  <tr>
+   <td>SUM
+   </td>
+   <td>19
+   </td>
+   <td>6
+   </td>
+   <td>25
+   </td>
+  </tr>
+  <tr>
+   <td>AVG
+   </td>
+   <td>84
+   </td>
+   <td>1
+   </td>
+   <td>85
+   </td>
+  </tr>
+  <tr>
+   <td>RATE
+   </td>
+   <td>0
+   </td>
+   <td>0
+   </td>
+   <td>0
+   </td>
+  </tr>
+  <tr>
+   <td>HISTOGRAM
+   </td>
+   <td>4
+   </td>
+   <td>11
+   </td>
+   <td>15
+   </td>
+  </tr>
+  <tr>
+   <td>
+   </td>
+   <td>
+   </td>
+   <td>Total
+   </td>
+   <td>138
+   </td>
+  </tr>
+</table>
+
+
+We can maintain a mapping of metric key (which can be anywhere between 30 to 50 characters) and metric value in the implementation of BaseStatsReporter class. Each of the above metric values is stored in a 8 byte integer type. At any moment we may have 138 × 8 = 1104 bytes of values data held in-memory. Also, we have 138 × ~50 = 6900 ~ 7K bytes for keys. In total we have 7K + 1104 ~ 8K bytes of memory. Thus holding metrics in memory has the least impact on Prestissimo workers.
```

Comment:
We can remove `Thus holding metrics in memory has the least impact on Prestissimo workers.`

Reply:
Done.